compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+        (~location: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  switch (comments) {
+  | [] => []
+  | [cmt, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(cmt);
+
+    let (_, cmtsline, cmtschar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtsline > stmt_end_line
+        || cmtsline == stmt_end_line
+        && cmtschar > stmt_end_char) {
+      [cmt, ...get_comments_after_location(~location, remaining_comments)];
+    } else {
+      get_comments_after_location(~location, remaining_comments);
+    };
+  };
+};
+
+let rec trailing_comments_inner =
+        (prev: option(Parsetree.comment), comments: list(Parsetree.comment)) => {
+  switch (prev) {
+  | None =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt] => [nobreak_comment_to_doc(cmt)]
+    | [cmt, ...rem] => [
+        comment_to_doc(cmt),
+        ...trailing_comments_inner(Some(cmt), rem),
+      ]
+    }
+  | Some(prev_cmt) =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt, ...rem] =>
+      let (_, prev_line, _, _) =
+        Locations.get_raw_pos_info(
+          Locations.get_comment_loc(prev_cmt).loc_end,
+        );
+      let (_, this_line, _, _) =
+        Locations.get_raw_pos_info(Locations.get_comment_loc(cmt).loc_start);
+
+      let comment_printer =
+        switch (rem) {
+        | [] => nobreak_comment_to_doc
+        | _ => comment_to_doc
+        };
+
+      switch (this_line - prev_line) {
+      | 0 => [
+          Doc.space,
+          comment_printer(cmt),
+          ...trailing_comments_inner(Some(cmt), rem),
+        ]
+
+      | 1 => [
+          switch (prev_cmt) {
+          | Line(_) => Doc.nil
+          | _ => Doc.hardLine
+          },
+          comment_printer(cmt),
+          ...trailing_comments_inner(Some(cmt), rem),
+        ]
+      | _ => [
+          Doc.hardLine,
+          comment_printer(cmt),
+          ...trailing_comments_inner(Some(cmt), rem),
+        ]
+      };
+    }
+  };
+};
+
+let block_trailing_comments_docs = (comments: list(Parsetree.comment)) =>
+  switch (comments) {
+  | [] => Doc.nil
+  | _remaining_comments =>
+    Doc.concat(trailing_comments_inner(None, comments))
+  };
+
+let single_line_of_comments = (comments: list(Parsetree.comment)) =>
+  switch (comments) {
+  | [] => Doc.nil
+  | _ =>
+    Doc.concat([
+      Doc.space,
+      Doc.join(
+        Doc.space,
+        List.map(c => {nobreak_comment_to_doc(c)}, comments),
+      ),
+    ])
+  };
+
+let rec new_comments_inner =
+        (prev: option(Parsetree.comment), comments: list(Parsetree.comment)) => {
+  switch (prev) {
+  | None =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt, ...rem] => [
+        comment_to_doc(cmt),
+        ...new_comments_inner(Some(cmt), rem),
+      ]
+    }
+  | Some(prev_cmt) =>
+    switch (comments) {
+    | [] => [Doc.nil]
```

Comment:
What happens if you do `Doc.concat([])`? Could this just be an empty list instead of including a `Doc.nil`?